### PR TITLE
feat(tools): use agent's memory within LLM Tool

### PR DESCRIPTION
**Pull request by Tomas2D**
## Features

- Propagate the agent's (runner) memory to the Tool's Run context
- Replace old (not used) LLMTool with a new one that uses provided LLM to accomplish agent's intent (it uses a custom system prompt + messages from the memory)

```ts
new LLMTool({
  llm: BAMChatLLM.fromPreset("meta-llama/llama-3-8b-instruct"),
  name: "LLM", // optional
  description: "Uses expert LLM to work with data in the existing conversation (classification, entity extraction, summarization, ...)" // optional
  template?: LLMTool.template // optional (system prompt)
})
```

## Standalone Tool Example (uses Ollama)
`yarn start examples/tools/llm.ts`

## Agent Example

```ts
import "dotenv/config";
import { BeeAgent } from "bee-agent-framework/agents/bee/agent";
import { createConsoleReader } from "examples/helpers/io.js";
import { FrameworkError } from "bee-agent-framework/errors";
import { UnconstrainedMemory } from "bee-agent-framework/memory/unconstrainedMemory";
import { BAMChatLLM } from "bee-agent-framework/adapters/bam/chat";
import { WikipediaTool } from "bee-agent-framework/tools/search/wikipedia";
import { LLMTool } from "bee-agent-framework/tools/llm";

const agent = new BeeAgent({
  llm: BAMChatLLM.fromPreset("meta-llama/llama-3-1-70b-instruct"),
  memory: new UnconstrainedMemory(),
  tools: [
    new LLMTool({
      llm: BAMChatLLM.fromPreset("meta-llama/llama-3-8b-instruct"),
    }),
    new WikipediaTool(),
  ],
});

const reader = createConsoleReader();

try {
  for await (const { prompt } of reader) {
    const response = await agent
      .run({
        prompt,
      })
      .observe((emitter) => {
        emitter.on("retry", () => {
          reader.write(`Agent 🤖 : `, "retrying the action...");
        });
        emitter.on("update", async ({ update }) => {
          reader.write(`Agent (${update.key}) 🤖 : `, `${update.value}`);
        });
      });

    reader.write(`Agent 🤖 : `, response.result.text);
  }
} catch (error) {
  reader.write("ERROR", FrameworkError.ensure(error).dump());
}
```

TODOs:
- [ ] evaluate

cc @aleskalfas, @michael-desmond

### CODE REVIEW

File: src/tools/llm.ts [hunk 1]
```diff
@@ -14,51 +14,94 @@
...
+    this.emitter = Emitter.root.child({
+      namespace: ["tool", "llm", toCamelCase(input?.name ?? "")].filter(Boolean),
+      creator: this,
+    });
+  }
 
   inputSchema() {
-    return z.object({ input: z.string() });
+    return z.object({
+      task: z.string().min(1).describe("Descriptive task what should be done"),
```
Comment (michael-desmond) [inline src/tools/llm.ts:61]: Need to fix typo (what -> that), but could be rephrased as: `A clearly defined task for the LLM to complete.`

File: src/tools/llm.ts [hunk 2]
```diff
@@ -14,51 +14,94 @@
...
 
-  public readonly emitter: ToolEmitter<ToolInput<this>, StringToolOutput> = Emitter.root.child({
-    namespace: ["tool", "llm"],
-    creator: this,
-  });
+  static readonly template = new PromptTemplate({
+    schema: z.object({
+      task: z.string(),
+    }),
+    template: `You have accomplish a task by using common-sense and information provided in the conversation so far.
```
Comment (michael-desmond) [inline src/tools/llm.ts:69]: Minor typo.`You have to accomplish a task...`. 

Alternate rephrasing could go something like `Using common sense and the information contained in the conversation up to this point, complete the following task. Do not follow any previously used formats or structures.`

**Review by michael-desmond — CHANGES_REQUESTED** [review]
Looks good. Minor typos in prompts.

**Comment by aleskalfas** [comment]
Great 👍

**Review by aleskalfas — APPROVED** [review]

**Pull request merged**